lib/int4range.ex

```diff
@@ -0,0 +1,102 @@
+defmodule Ecto.Int4Range do
+  @moduledoc """
+  A Postgres range of `int4` integers. Equivalent to `int4range`.
+  """
+
+  use Ecto.Type
+
+  @int4_range -2_147_483_648..2_147_483_647
+
+  @impl Ecto.Type
+  def type, do: :int4range
+
+  @impl Ecto.Type
+  def cast(%Postgrex.Range{} = range) do
+    to_postgrex_range(range)
+  end
+
+  def cast({lower, upper}) do
+    to_postgrex_range({lower, upper})
+  end
+
+  def cast(_), do: :error
+
+  @impl Ecto.Type
+  def dump(%Postgrex.Range{} = range) do
+    {:ok, range}
+  end
+
+  def dump(_), do: :error
+
+  @impl Ecto.Type
+  def load(%Postgrex.Range{} = range) do
+    {:ok, normalize_range(range)}
+  end
+
+  @doc """
+  Checks and converts a `Postgrex.Range` or tuple into a `Postgrex.Range.t()`
+
+  ## Examples
+
+      iex> Ecto.DateRange.to_postgrex_range({1, 3})
+      %Postgrex.Range{lower: 1, upper: 3, lower_inclusive: true, upper_inclusive: true}
+
+  """
+  @spec to_postgrex_range(Postgrex.Range.t() | {integer(), integer()}) ::
+          Postgrex.Range.t() | :error
+  def to_postgrex_range({lower, upper}) do
```

Comment:
@PJUllrich this looks great overall, my only "issue" is that this function is that it has a different API than https://github.com/bitfo/ecto_date_range/blob/main/lib/ecto_date_range.ex#L70

It's also doing conversion AND validation, instead of just conversion. IMO it should only convert and leave validation to the Ecto callbacks. If a user supplies bad data to `to_postgrex_range()` we should not try to raise or return an error

Reply:
Good catch! (I shouldn't code with tired eyes). I fixed it :)

Reply:
Regarding the validation: I tried to move the validation to a single place instead of using similar guards in both type callbacks. But in this case, I agree that it's better to have some duplication if it keeps the helper functions "pure" 👍